runtime/src/main/java/org/corfudb/runtime/view/stream/BackpointerStreamView.java

```diff
@@ -0,0 +1,367 @@
+                    last.contextID, last.maxAddress);
+        }
+
+        // If we have no entries to read, fill the read queue.
+        if (getCurrentContext().readQueue.isEmpty()) {
+            fillReadQueue(maxGlobal);
+            // If we still have no entries to read, there's nothing in
+            // the stream (we linearized the read during fillReadQueue()).
+            if (getCurrentContext().readQueue.isEmpty()) {
+                return null;
+            }
+        }
+
+        // If the lowest value in the read queue is greater than
+        // max global, there's nothing for us to do.
+        if (getCurrentContext().readQueue.first() > maxGlobal) {
+            return null;
+        }
+
+        // Otherwise we remove entries one at a time from the read queue,
+        // reading each entry in order until one of the entries contains
+        // data -or- we exceed maxGlobal -or- there are no more entries. If
+        // we exceed maxGlobal or run out of entries, we return null, else
+        // we return the data.;
+
+        while ( getCurrentContext().readQueue.size() > 0 &&
+                getCurrentContext().readQueue.first() <= maxGlobal) {
+            final long thisRead = getCurrentContext().readQueue.pollFirst();
+            getCurrentContext().logPointer = thisRead;
+            LogData ld = runtime.getAddressSpaceView().read(thisRead);
+            if (ld.containsStream(getCurrentContext().contextID)) {
+                // If this is a copy-on-write entry, add a new stream
+                // context.
+                Object res = ld.getPayload(runtime);
+                if (res instanceof StreamCOWEntry) {
+                    StreamCOWEntry ce = (StreamCOWEntry) res;
+                    streamContexts
+                            .add(new StreamContext(ce.getOriginalStream(),
+                                    ce.getFollowUntil()));
+                    // To follow the underlying stream, we'll call
+                    // nextUpTo again.
+                    return nextUpTo(maxGlobal);
+                } else {
+                    // Otherwise return the log data.
+                    return ld;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * In the backpointer based implementation, we indicate we may have
+     * entries available if the read queue contains entries to read -or-
+     * if the next token is greater than our log pointer.
+     */
+    @Override
+    public boolean hasNext() {
+        return  !getCurrentContext().readQueue.isEmpty() ||
+                runtime.getSequencerView()
+                .nextToken(Collections.singleton(getCurrentContext()
+                        .contextID), 0).getToken()
+                        > getCurrentContext().logPointer;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void close() {}
+
+
+    /**
+     * Fill the read queue for the current context.
```

Comment:
explain what you mean by a "read queue" : a queue containing updatePositions (Adresses, ahem), which contains all the forward positions for this stream.

Note that, the queue may potentially include irrelevant positions, which we will discover when we actually read them. For example, some of these address may be holes. Or. these may be positions that we single-step through, in section of the log that has no backpointers information.